t/cmd/lfstest-count-tests.go

```diff
@@ -0,0 +1,270 @@
+				fmt.Sprintf("LFSTEST_URL=%s", os.Getenv("LFSTEST_URL")),
+				fmt.Sprintf("LFSTEST_SSL_URL=%s", os.Getenv("LFSTEST_SSL_URL")),
+				fmt.Sprintf("LFSTEST_CLIENT_CERT_URL=%s", os.Getenv("LFSTEST_CLIENT_CERT_URL")),
+				fmt.Sprintf("LFSTEST_DIR=%s", os.Getenv("LFSTEST_DIR")),
+				fmt.Sprintf("LFSTEST_CERT=%s", os.Getenv("LFSTEST_CERT")),
+				fmt.Sprintf("LFSTEST_CLIENT_CERT=%s", os.Getenv("LFSTEST_CLIENT_CERT")),
+				fmt.Sprintf("LFSTEST_CLIENT_KEY=%s", os.Getenv("LFSTEST_CLIENT_KEY")),
+			)
+			cmd.Stdout = log
+
+			// Start performs a fork/execve, hence we can abandon
+			// this process once it has started.
+			if err := cmd.Start(); err != nil {
+				return n, err
+			}
+			return n + 1, nil
+		})
+	case "decrement":
+		err = callWithCount(func(n int) (int, error) {
+			if n > 1 {
+				// If there is at least two tests running, we
+				// need not shutdown a lfstest-gitserver
+				// instance.
+				return n - 1, nil
+			}
+
+			// Otherwise, we need to POST to /shutdown, which will
+			// cause the lfstest-gitserver to abort itself.
+			url, err := ioutil.ReadFile(os.Getenv("LFS_URL_FILE"))
+			if err == nil {
+				_, err = http.Post(string(url)+"/shutdown",
+					"application/text",
+					strings.NewReader(time.Now().String()))
+			}
+
+			return 0, nil
+		})
+	}
+
+	if err != nil {
+		fatal(err)
+	}
+}
+
+var (
+	// acquireTick is the constant time that one tick (i.e., one attempt at
+	// acquiring the lock) should last.
+	acquireTick = 10 * time.Millisecond
+)
+
+// acquire acquires the lock file necessary to perform updates to test_count,
+// and returns an error if that lock cannot be acquired.
+func acquire(ctx context.Context) error {
+	path, err := path(lockFile)
+	if err != nil {
+		return err
+	}
+
+	tick := time.NewTicker(acquireTick)
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-tick.C:
+			// Try every tick of the above ticker before giving up
+			// and trying again.
+			_, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL, 0666)
+			if err == nil || !os.IsExist(err) {
+				return err
+			}
+		case <-ctx.Done():
+			// If the context.Context above has reached its
+			// deadline, we must give up.
+			return errCouldNotAcquire
+		}
+	}
+}
+
+// release releases the lock file so that another process can take over, or
+// returns an error.
+func release() error {
+	path, err := path(lockFile)
+	if err != nil {
+		return err
+	}
+	return os.Remove(path)
+}
+
+// callWithCount calls the given countFn with the current count in test_count,
+// and updates it with what the function returns.
+//
+// If the function produced an error, that will be returned instead.
+func callWithCount(fn countFn) error {
+	path, err := path(countFile)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0666)
+	if err != nil {
+		return err
+	}
+
+	contents, err := ioutil.ReadAll(f)
+	if err != nil {
+		return err
+	}
+
+	var n int = 0
+	if len(contents) != 0 {
+		n, err = strconv.Atoi(string(contents))
```

Comment:
Would it be worth erroring out here if the `n` is less than 0? It seems like a good safety to have since the number is being read from a file.

Reply:
Great idea! I got that as: 8f6da107 (t/cmd/lfstest-count-tests.go: return errNegativeCount when negative, 2018-07-16).